lance-spark-3.5_2.12/src/main/java/com/lancedb/lance/spark/write/SparkPositionDeltaWrite.java

```diff
@@ -0,0 +1,236 @@
+    return Distributions.clustered(new NamedReference[] {segmentId});
+  }
+
+  @Override
+  public SortOrder[] requiredOrdering() {
+    NamedReference segmentId = Expressions.column(LanceConstant.ROW_ADDRESS);
+    SortValue sortValue =
+        new SortValue(segmentId, SortDirection.ASCENDING, NullOrdering.NULLS_FIRST);
+    return new SortValue[] {sortValue};
+  }
+
+  @Override
+  public DeltaBatchWrite toBatch() {
+    return new PositionDeltaBatchWrite();
+  }
+
+  private class PositionDeltaBatchWrite implements DeltaBatchWrite {
+
+    @Override
+    public DeltaWriterFactory createBatchWriterFactory(PhysicalWriteInfo info) {
+      return new PositionDeltaWriteFactory(sparkSchema, config);
+    }
+
+    @Override
+    public void commit(WriterCommitMessage[] messages) {
+      List<Long> removedFragmentIds = new ArrayList<>();
+      List<FragmentMetadata> updatedFragments = new ArrayList<>();
+      List<FragmentMetadata> newFragments = new ArrayList<>();
+
+      Arrays.stream(messages)
+          .map(m -> (DeltaWriteTaskCommit) m)
+          .forEach(
+              m -> {
+                removedFragmentIds.addAll(m.removedFragmentIds());
+                updatedFragments.addAll(m.updatedFragments());
+                newFragments.addAll(m.newFragments());
+              });
+
+      LanceDatasetAdapter.updateFragments(
+          config, removedFragmentIds, updatedFragments, newFragments);
+    }
+
+    @Override
+    public void abort(WriterCommitMessage[] messages) {}
+  }
+
+  private static class PositionDeltaWriteFactory implements DeltaWriterFactory {
+    private final StructType sparkSchema;
+    private final LanceConfig config;
+
+    PositionDeltaWriteFactory(StructType sparkSchema, LanceConfig config) {
+      this.sparkSchema = sparkSchema;
+      this.config = config;
+    }
+
+    @Override
+    public DeltaWriter<InternalRow> createWriter(int partitionId, long taskId) {
+      int batch_size = SparkOptions.getBatchSize(config);
+      LanceArrowWriter arrowWriter = LanceDatasetAdapter.getArrowWriter(sparkSchema, batch_size);
+      WriteParams params = SparkOptions.genWriteParamsFromConfig(config);
+      Callable<List<FragmentMetadata>> fragmentCreator =
+          () -> LanceDatasetAdapter.createFragment(config.getDatasetUri(), arrowWriter, params);
+      FutureTask<List<FragmentMetadata>> fragmentCreationTask = new FutureTask<>(fragmentCreator);
+      Thread fragmentCreationThread = new Thread(fragmentCreationTask);
+      fragmentCreationThread.start();
+
+      return new LanceDeltaWriter(
+          config, new LanceDataWriter(arrowWriter, fragmentCreationTask, fragmentCreationThread));
+    }
+  }
+
+  private static class LanceDeltaWriter implements DeltaWriter<InternalRow> {
+    private final LanceConfig config;
+    private final LanceDataWriter writer;
+
+    // Key is fragmentId, Value is fragment's deleted row indexes
+    private final Map<Integer, List<Integer>> deletedRows;
```

Comment:
We can use a `TreeMap<Integer, RoaringBitmap>` for better efficiency (basically equivalent to `RoaringTreemap` in rust.